stan/math/rev/mat/fun/matrix_power.hpp

```diff
@@ -0,0 +1,182 @@
+#ifndef STAN_MATH_REV_MAT_FUN_MATRIX_POWER_HPP
+#define STAN_MATH_REV_MAT_FUN_MATRIX_POWER_HPP
+
+#include <stan/math/prim/scal/err/invalid_argument.hpp>
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/mat/fun/typedefs.hpp>
+#include <stan/math/prim/mat/err/check_finite.hpp>
+#include <stan/math/prim/mat/err/check_square.hpp>
+#include <stan/math/rev/core.hpp>
+#include <stan/math/rev/scal/fun/value_of.hpp>
+#include <stan/math/rev/mat/fun/typedefs.hpp>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+namespace internal {
+
+template <int R, int C>
+class matrix_product_vari_n0 : public vari {
+ public:
+  double* Mn_;
+  vari** adjMnRef_;
+
+  explicit matrix_product_vari_n0(const Eigen::Matrix<var, R, C>& M)
+      : vari(0.0),
+        Mn_(reinterpret_cast<double*>(
```

Comment:
There's an alloc-array method that'll avoid the casts and sizeof.

Reply:
So there is. Code looks a lot better now. Thank you.